cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,425 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#ifndef KVIKIO_LIBCURL_FOUND
+#error "cannot include <kvikio/remote_handle.hpp>, configuration did not find libcurl"
+#endif
+
+#include <cstddef>
+#include <cstring>
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <curl/curl.h>
+
+#include <kvikio/defaults.hpp>
+#include <kvikio/error.hpp>
+#include <kvikio/parallel_operation.hpp>
+#include <kvikio/posix_io.hpp>
+#include <kvikio/utils.hpp>
+
+namespace kvikio {
+namespace detail {
+
+/**
+ * @brief Singleton class to initialize and cleanup the global state of libcurl
+ *
+ * https://curl.se/libcurl/c/libcurl.html
+ * In a C++ module, it is common to deal with the global constant situation by defining a special
+ * class that represents the global constant environment of the module. A program always has exactly
+ * one object of the class, in static storage. That way, the program automatically calls the
+ * constructor of the object as the program starts up and the destructor as it terminates. As the
+ * author of this libcurl-using module, you can make the constructor call curl_global_init and the
+ * destructor call curl_global_cleanup and satisfy libcurl's requirements without your user having
+ * to think about it. (Caveat: If you are initializing libcurl from a Windows DLL you should not
+ * initialize it from DllMain or a static initializer because Windows holds the loader lock during
+ * that time and it could cause a deadlock.)
+ */
+class LibCurl {
+ public:
+  // We hold an unique pointer to the raw curl handle and sets `curl_easy_cleanup` as its Deleter.
+  using UniqueHandlePtr = std::unique_ptr<CURL, std::function<void(CURL*)>>;
+
+ private:
+  std::mutex _mutex{};
+  // Curl handles free to be used.
+  std::vector<UniqueHandlePtr> _free_curl_handles{};
+
+  LibCurl()
+  {
+    CURLcode err = curl_global_init(CURL_GLOBAL_DEFAULT);
+    if (err != CURLE_OK) {
+      throw std::runtime_error("cannot initialize libcurl - errorcode: " + std::to_string(err));
+    }
+    curl_version_info_data* ver = curl_version_info(::CURLVERSION_NOW);
+    if ((ver->features & CURL_VERSION_THREADSAFE) == 0) {
+      throw std::runtime_error("cannot initialize libcurl - built with thread safety disabled");
+    }
+  }
+  ~LibCurl() noexcept
+  {
+    _free_curl_handles.clear();
+    curl_global_cleanup();
+  }
+
+  LibCurl(LibCurl const&)            = delete;
+  LibCurl& operator=(LibCurl const&) = delete;
+  LibCurl(LibCurl&& o)               = delete;
+  LibCurl& operator=(LibCurl&& o)    = delete;
+
+ public:
+  static LibCurl& instance()
+  {
+    static LibCurl _instance;
```

Comment:
*question/thought*: Since this is an object with static lifetime, the destructor will run below main. I had a read through the libcurl documentation and did not manage to determine if this is safe. They don't explicitly call it out as unsafe, so maybe it's fine?

Reply:
I think this is what they are saying in the docs? From https://curl.se/libcurl/c/libcurl.html:

> In a C++ module, it is common to deal with the global constant situation by defining a special class that represents the global constant environment of the module. A program always has exactly one object of the class, in static storage. That way, the program automatically calls the constructor of the object as the program starts up and the destructor as it terminates. As the author of this libcurl-using module, you can make the constructor call [curl_global_init](https://curl.se/libcurl/c/curl_global_init.html) and the destructor call [curl_global_cleanup](https://curl.se/libcurl/c/curl_global_cleanup.html) and satisfy libcurl's requirements without your user having to think about it. (Caveat: If you are initializing libcurl from a Windows DLL you should not initialize it from DllMain or a static initializer because Windows holds the loader lock during that time and it could cause a deadlock.)

Reply:
Thanks, I had a look at the code too and there's nothing obviously there that will bite, AFAICT (curl only registers an `atexit` callback on amigaos)